File: HighwayPursuit/src/entities/weapons.py
"""
Armas do ladrão (etapa 9): óleo e tiro traseiro.

Diferente do spike strip e da barreira, essas armas não têm aviso --
é o ladrão surpreendendo a polícia, não o contrário.
"""

import pygame
from src import settings


class OleoDerrapante:
    """
    Mancha de óleo largada no chão. Comporta-se como um objeto fixo no
    mundo (rola na velocidade da pista), igual ao spike strip e à
    barreira. Só afeta a polícia, e só uma vez (soma usada = True).
    """

    def __init__(self, faixa, y_inicial):
        self.faixa = faixa
        self.usada = False

        self.x = settings.MARGEM_PISTA + faixa * settings.LARGURA_FAIXA + settings.LARGURA_FAIXA / 2
        self.y = y_inicial

        self.rect = pygame.Rect(0, 0, settings.LARGURA_FAIXA - 30, 34)
        self.rect.center = (int(self.x), int(self.y))

    def atualizar(self):
        self.y += settings.VELOCIDADE_PISTA
        self.rect.center = (int(self.x), int(self.y))

    def fora_da_tela(self):
        return self.rect.top > settings.ALTURA_TELA

    def desenhar(self, tela):
        pygame.draw.ellipse(tela, settings.COR_OLEO, self.rect)
        brilho = pygame.Rect(0, 0, self.rect.width * 0.4, self.rect.height * 0.35)
        brilho.center = (self.rect.centerx - 6, self.rect.centery - 4)
        pygame.draw.ellipse(tela, settings.COR_OLEO_BRILHO, brilho)


class Projetil:
    """
    Tiro disparado pelo ladrão para trás. Viaja numa faixa fixa (a
    faixa do ladrão no momento do disparo) e some ao acertar a
    polícia ou sair da tela.
    """

    def __init__(self, faixa, y_inicial):
        self.faixa = faixa
        self.atingiu = False

        self.x = settings.MARGEM_PISTA + faixa * settings.LARGURA_FAIXA + settings.LARGURA_FAIXA / 2
        self.y = y_inicial

        self.rect = pygame.Rect(0, 0, 10, 22)
        self.rect.center = (int(self.x), int(self.y))

    def atualizar(self):
        self.y += settings.VELOCIDADE_PROJETIL
        self.rect.center = (int(self.x), int(self.y))

    def fora_da_tela(self):
        return self.rect.top > settings.ALTURA_TELA

    def desenhar(self, tela):
        pygame.draw.rect(tela, settings.COR_PROJETIL, self.rect, border_radius=4)

File: HighwayPursuit/web/src/entities/Weapons.ts
import * as S from "../settings";
import { Rect } from "../utils/rect";

export class OleoDerrapante {
  faixa: number;
  usada = false;
  x: number;
  y: number;
  rect: Rect;

  constructor(faixa: number, yInicial: number) {
    this.faixa = faixa;
    this.x = S.MARGEM_PISTA + faixa * S.LARGURA_FAIXA + S.LARGURA_FAIXA / 2;
    this.y = yInicial;
    this.rect = new Rect(0, 0, S.LARGURA_FAIXA - 30, 34);
    this.rect.center = [this.x, this.y];
  }

  atualizar() {
    this.y += S.VELOCIDADE_PISTA;
    this.rect.center = [Math.round(this.x), Math.round(this.y)];
  }

  foraDaTela() {
    return this.rect.top > S.ALTURA_TELA;
  }

  desenhar(ctx: CanvasRenderingContext2D) {
    ctx.fillStyle = S.rgb(S.COR_OLEO);
    ctx.beginPath();
    ctx.ellipse(this.x, this.y, this.rect.width / 2, this.rect.height / 2, 0, 0, Math.PI * 2);
    ctx.fill();
    ctx.fillStyle = S.rgb(S.COR_OLEO_BRILHO);
    ctx.beginPath();
    ctx.ellipse(this.x - 6, this.y - 4, this.rect.width * 0.2, this.rect.height * 0.175, 0, 0, Math.PI * 2);
    ctx.fill();
  }
}

export class Projetil {
  faixa: number;
  atingiu = false;
  x: number;
  y: number;
  rect: Rect;

  constructor(faixa: number, yInicial: number) {
    this.faixa = faixa;
    this.x = S.MARGEM_PISTA + faixa * S.LARGURA_FAIXA + S.LARGURA_FAIXA / 2;
    this.y = yInicial;
    this.rect = new Rect(0, 0, 10, 22);
    this.rect.center = [this.x, this.y];
  }

  atualizar() {
    this.y += S.VELOCIDADE_PROJETIL;
    this.rect.center = [Math.round(this.x), Math.round(this.y)];
  }

  foraDaTela() {
    return this.rect.top > S.ALTURA_TELA;
  }

  desenhar(ctx: CanvasRenderingContext2D) {
    ctx.fillStyle = S.rgb(S.COR_PROJETIL);
    ctx.fillRect(this.rect.left, this.rect.top, this.rect.width, this.rect.height);
  }
}
